pkg/server/container_stats_list.go

```diff
@@ -17,14 +17,127 @@ limitations under the License.
 package server
 
 import (
-	"errors"
-
+	"fmt"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/cgroups"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/typeurl"
 	"k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
 )
 
 // ListContainerStats returns stats of all running containers.
 func (c *criContainerdService) ListContainerStats(ctx context.Context, in *runtime.ListContainerStatsRequest) (*runtime.ListContainerStatsResponse, error) {
-	return nil, errors.New("not implemented")
+	request, err := c.buildTaskMetricsRequest(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build metrics request: %v", err)
+	}
+	resp, err := c.taskService.Metrics(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch metrics for tasks: %v", err)
+	}
+	criStats, err := c.toCRIContainerStats(resp.Metrics)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to cri containerd stats format: %v", err)
+	}
+	return criStats, nil
+}
+
+func (c *criContainerdService) toCRIContainerStats(stats []*types.Metric) (*runtime.ListContainerStatsResponse, error) {
+	containerStats := new(runtime.ListContainerStatsResponse)
+	for _, stat := range stats {
+		var cs runtime.ContainerStats
+		if err := c.getContainerMetrics(stat, &cs); err != nil {
+			return nil, fmt.Errorf("failed to decode container metrics: %v", err)
+		}
+		containerStats.Stats = append(containerStats.Stats, &cs)
+	}
+	return containerStats, nil
+}
+
+func (c *criContainerdService) getContainerMetrics(stats *types.Metric, cs *runtime.ContainerStats) error {
+	s, err := typeurl.UnmarshalAny(stats.Data)
+	if err != nil {
+		return fmt.Errorf("failed to extract container metrics: %v", err)
+	}
+	metrics := s.(*cgroups.Metrics)
+	cs.Cpu = &runtime.CpuUsage{
+		Timestamp:            stats.Timestamp.Unix(),
+		UsageCoreNanoSeconds: &runtime.UInt64Value{metrics.CPU.Usage.Total},
+	}
+	cs.Memory = &runtime.MemoryUsage{
+		Timestamp:       stats.Timestamp.Unix(),
+		WorkingSetBytes: &runtime.UInt64Value{metrics.Memory.Usage.Usage},
+	}
+
+	var usedBytes, inodesUsed uint64
+	sn, err := c.snapshotStore.Get(stats.ID)
+	// If snapshotstore doesnt have cached snapshotStore
+	// set WritableLayer to zero
+	if err == nil {
+		inodesUsed = sn.Size
+		usedBytes = sn.Inodes
+	}
+	cs.WritableLayer = &runtime.FilesystemUsage{
+		Timestamp: stats.Timestamp.Unix(),
+		StorageId: &runtime.StorageIdentifier{
+			Uuid: c.imageFSUUID,
+		},
+		UsedBytes:  &runtime.UInt64Value{usedBytes},
+		InodesUsed: &runtime.UInt64Value{inodesUsed},
+	}
+
+	// Get the container from store and extract the attributes
+	cnt, err := c.containerStore.Get(stats.ID)
```

Comment:
It's possible that a container shows up during list, but get deleted before this point. In that case, we should not return error for the whole `ListContainerStats`.

We should either pass in the container config into this function, or distinguish not found error here.